### PR TITLE
ENG-13834: distributed sorts with merge receive

### PR DESCRIPTION
**Pull request by wolffcm**

### CODE REVIEW

File: src/frontend/org/voltdb/planner/microoptimizations/InlineOrderByIntoMergeReceive.java [hunk 1]
```diff
@@ -316,4 +336,10 @@ AbstractPlanNode convertToSerialAggregation(AbstractPlanNode aggregateNode, Orde
         return aggregateNode;
     }
 
+    // For the time being coordinator fragment is considered to be trivial if it consists of
+    // only RECEIVE plan node
+    private boolean hasTrivialCoordinatorFragment(AbstractPlanNode planNode) {
```
Comment (wolffcm) [inline src/frontend/org/voltdb/planner/microoptimizations/InlineOrderByIntoMergeReceive.java]: Actually the pattern you care about is 
```
...
  OrderBy
    Receive
      Send
         ...
```
Right? There could still be other stuff above the OrderBy node, and it would still match the pattern?

Comment (wolffcm) [reply src/frontend/org/voltdb/planner/microoptimizations/InlineOrderByIntoMergeReceive.java]: I think this class already looks for a similar pattern, I just think this method's name isn't quite right.

File: src/frontend/org/voltdb/plannodes/OrderByPlanNode.java [hunk 2]
```diff
@@ -79,6 +79,18 @@ public void validate() throws Exception {
         }
     }
 
+    /**
+     * Add multiple sort to the order-by
+     * @param sortExprs  List of the input expression on which to order the rows
+     * @param sortDirs List of the corresponding sort order for each input expression
+     */
+    public void addSort(List<AbstractExpression> sortExprs, List<SortDirectionType> sortDirs) {
```
Comment (wolffcm) [inline src/frontend/org/voltdb/plannodes/OrderByPlanNode.java]: This is a poorly-named method.  The method you are overloading is poorly-named too.  Maybe rename it to `addSortKey`?

File: tests/frontend/org/voltdb/planner/TestPlansGroupBy.java [hunk 3]
```diff
@@ -712,11 +713,12 @@ public void testEdgeComplexRelatedCases() {
         pns = compileToFragments("select PKEY+A1 from T1 Order by PKEY+A1");
         AbstractPlanNode p = pns.get(0).getChild(0);
         assertTrue(p instanceof ProjectionPlanNode);
-        assertTrue(p.getChild(0) instanceof OrderByPlanNode);
-        assertTrue(p.getChild(0).getChild(0) instanceof ReceivePlanNode);
+        assertTrue(p.getChild(0) instanceof MergeReceivePlanNode);
```
Comment (wolffcm) [inline tests/frontend/org/voltdb/planner/TestPlansGroupBy.java]: Nice!!

File: tests/frontend/org/voltdb/planner/TestPlansOrderBy.java [hunk 4]
```diff
@@ -361,7 +361,7 @@ public void testOrderByMPOptimized() {
         }
         {
             // P_D1_IDX index provides the right order for the coordinator. Merge Receive with LIMIT
-            validateMergeReceive("select P_D1 from P order by P_D1 limit 3", true, new int[] {0});
+            validateMergeReceive("select P_D1 from P order by P_D1 limit 3 offset 1", true, new int[] {0});
```
Comment (wolffcm) [inline tests/frontend/org/voltdb/planner/TestPlansOrderBy.java]: Why did you add the offset here?  It is a good test case, I'm just curious.

File: tests/frontend/org/voltdb/planner/TestPlansOrderBy.java [hunk 5]
```diff
@@ -390,54 +388,54 @@ public void testOrderByMPOptimized() {
...
         {
             // P_D2 column is not covered by P_D1_IDX index. ORDER BY
             List<AbstractPlanNode> frags =  compileToFragments(
-                    "select P_D1 from P where P_D1 > 1 order by P_D1, P_D2");
+                    "select P_D1 from P where P_D1 > 1 order by abs(P_D3), P_D2");
             assertEquals(2, frags.size());
             AbstractPlanNode pn = frags.get(0).getChild(0).getChild(0);
-            assertEquals(PlanNodeType.ORDERBY, pn.getPlanNodeType());
+            assertEquals(PlanNodeType.MERGERECEIVE, pn.getPlanNodeType());
+            assertNotNull(pn.getInlinePlanNode(PlanNodeType.ORDERBY));
```
Comment (wolffcm) [inline tests/frontend/org/voltdb/planner/TestPlansOrderBy.java]: You should verify that the order by node appears in collector fragment below the send node, here and also for the other test cases.

Comment (wolffcm) [reply tests/frontend/org/voltdb/planner/TestPlansOrderBy.java]: Can you call `validateMergeReceive` for these?  Also can you make `validateMergeReceive` verify that there is a ORDER BY node in the collector fragment if it's needed (perhaps by adding an argument to `validateMergeReceive`)

File: tests/frontend/org/voltdb/planner/TestWindowedFunctions.java [hunk 6]
```diff
@@ -777,14 +777,14 @@ public void testWindowFunctionsWithIndexes() {
                                   PlanNodeType.SEQSCAN));
         }
         //  12: SLOB, No WF, MP Query, noindex
-        //      Expect OrderBy(SLOB) -> RECV -> SEND -> SeqScan
+        //      Expect MergeReceive with OrderBy(SLOB) -> RECV -> SEND -> ORDER BY -> SeqScan
```
Comment (wolffcm) [inline tests/frontend/org/voltdb/planner/TestWindowedFunctions.java]: Nice, so this optimization will apply to window functions in the OVER clause too...

File: tests/frontend/org/voltdb/planner/TestPlansGroupBy.java (hunk 3 above)
Comment (BillWhite) [inline tests/frontend/org/voltdb/planner/TestPlansGroupBy.java]: Is it possible that for some queries the projection node will be removed, because it is just the identity function?

File: src/frontend/org/voltdb/planner/microoptimizations/InlineOrderByIntoMergeReceive.java (hunk 1 above)
Comment (alexeevm) [reply src/frontend/org/voltdb/planner/microoptimizations/InlineOrderByIntoMergeReceive.java]: Agree. The name is a bit misleading. I renamed it to be isOptimizationPossible and moved all the checks there.

File: src/frontend/org/voltdb/plannodes/OrderByPlanNode.java (hunk 2 above)
Comment (alexeevm) [reply src/frontend/org/voltdb/plannodes/OrderByPlanNode.java]: Renamed them to be addSortExpression and addSortExpressions

File: tests/frontend/org/voltdb/planner/TestPlansOrderBy.java (hunk 4 above)
Comment (alexeevm) [reply tests/frontend/org/voltdb/planner/TestPlansOrderBy.java]: I just wanted to be sure that PlanAssembler does push down a Limit node with both limit and offset set as part of the regular optimization. This goes back to the comment I added to the InlineOrderByIntoMergeReceive - if PlanAssembler fails to push down the Limit node, the MERGE RECEIVE is not possible.

File: tests/frontend/org/voltdb/planner/TestPlansOrderBy.java (hunk 5 above)
Comment (alexeevm) [reply tests/frontend/org/voltdb/planner/TestPlansOrderBy.java]: Done

File: tests/frontend/org/voltdb/planner/TestPlansGroupBy.java (hunk 3 above)
Comment (alexeevm) [reply tests/frontend/org/voltdb/planner/TestPlansGroupBy.java]: Currently the optimization does not remove any projections.
But in this particular case you do need a projections to compute the PKEY+A1 expression, right?

**Comment by alexeevm** [comment]
I pulled the latest master from the VoltDB repo and rerunning the tests

File: tests/frontend/org/voltdb/planner/TestPlansOrderBy.java (hunk 5 above)
Comment (wolffcm) [reply tests/frontend/org/voltdb/planner/TestPlansOrderBy.java]: You said "done," but I don't see those changes...  did you forget to push a commit up to GitHub?

Comment (alexeevm) [reply tests/frontend/org/voltdb/planner/TestPlansOrderBy.java]: I haven't pushed the changes yet because I was validating the unit results. I see three test cases failed
     [exec] See obj/release/testoutput/report/junit-noframes.html for details.
     [exec] Suite TestFuzzMeshArbiter reports 1 failure(s).
     [exec]   Testcase testFuzz junit.framework.AssertionFailedError
     [exec] Suite TestFunctionsForVoltDBSuite reports 2 failure(s).
     [exec]   Testcase testInternetAddresses-localCluster-1-1-JNI junit.framework.AssertionFailedError
     [exec]   Testcase testInternetAddresses-localCluster-3-1-JNI junit.framework.AssertionFailedError

I don't think the failures are related to my changes because they are reproducible without them as well.

Just pushed the updates

**Comment by wolffcm** [comment]
I merged this on a separate branch.
